presto-main/src/main/java/com/facebook/presto/sql/planner/sanity/VerifyNoOriginalExpression.java

```diff
@@ -0,0 +1,167 @@
+import com.facebook.presto.spi.plan.PlanNode;
+import com.facebook.presto.spi.plan.ValuesNode;
+import com.facebook.presto.spi.relation.RowExpression;
+import com.facebook.presto.sql.parser.SqlParser;
+import com.facebook.presto.sql.planner.SimplePlanVisitor;
+import com.facebook.presto.sql.planner.TypeProvider;
+import com.facebook.presto.sql.planner.plan.AggregationNode;
+import com.facebook.presto.sql.planner.plan.ApplyNode;
+import com.facebook.presto.sql.planner.plan.JoinNode;
+import com.facebook.presto.sql.planner.plan.ProjectNode;
+import com.facebook.presto.sql.planner.plan.SpatialJoinNode;
+import com.facebook.presto.sql.planner.plan.StatisticAggregations;
+import com.facebook.presto.sql.planner.plan.TableFinishNode;
+import com.facebook.presto.sql.planner.plan.TableWriterNode;
+import com.facebook.presto.sql.planner.plan.WindowNode;
+
+import java.util.Optional;
+
+import static com.facebook.presto.sql.relational.OriginalExpressionUtils.isExpression;
+import static com.google.common.base.Preconditions.checkArgument;
+
+/**
+ * Ensures that FilterNode and ValuesNode have no OriginalExpression left
+ */
+public class VerifyNoOriginalExpression
+        implements PlanSanityChecker.Checker
+{
+    @Override
+    public void validate(PlanNode plan, Session session, Metadata metadata, SqlParser sqlParser, TypeProvider types, WarningCollector warningCollector)
+    {
+        plan.accept(new Visitor(), null);
+    }
+
+    private static class Visitor
+            extends SimplePlanVisitor<Void>
+    {
+        @Override
+        public Void visitProject(ProjectNode node, Void context)
+        {
+            visitPlan(node, context);
+
+            node.getAssignments().getExpressions().forEach(
+                    expression -> checkNotOriginalExpression(expression));
+            return null;
+        }
+
+        @Override
+        public Void visitAggregation(AggregationNode node, Void context)
+        {
+            visitPlan(node, context);
+
+            node.getAggregations().values().forEach(Visitor::checkAggregation);
+            return null;
+        }
+
+        @Override
+        public Void visitWindow(WindowNode node, Void context)
+        {
+            visitPlan(node, context);
+
+            node.getWindowFunctions().values().forEach(
+                    windowModeFunction -> windowModeFunction.getFunctionCall().getArguments().forEach(expression -> checkNotOriginalExpression(expression)));
+            return null;
+        }
+
+        @Override
+        public Void visitFilter(FilterNode node, Void context)
+        {
+            visitPlan(node, context);
+
+            checkNotOriginalExpression(node.getPredicate());
+            return null;
+        }
+
+        @Override
+        public Void visitValues(ValuesNode node, Void context)
+        {
+            visitPlan(node, context);
+
+            node.getRows().forEach(
+                    rows -> rows.forEach(row -> checkNotOriginalExpression(row)));
```

Comment:
you can probably do `rows.forEach(Visitor::checkNotOriginalExpression)`. As well as the other call sites.

Reply:
> Visitor::checkNotOriginalExpression

great advice! Thanks!